### PR TITLE
Prevent complete animators from causing drawing

### PR DESCRIPTION
**Pull request by macdrevx**
Fixes: #469

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Update changelog

### Summary of changes

- No public API changes
- Fixed an issue where animators created by fly to and ease to were not released until the next fly to or ease to began
- Fixed an issue where a complete animator would trigger redrawing unnecessarily
- Refactored related code for improved testability
- Reorganized files in the MapView and MapViewTests directories into Foundation and FoundationTests, respectively

### CODE REVIEW

**Comment by macdrevx** [comment]
Still need to update the project file

**Comment by nishant-karajgikar** [comment]
Could we run device tests before merging this PR? You can use `[run device tests]`.

File: Sources/MapboxMaps/Foundation/MapViewDependencyProvider.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+internal final class MapViewDependencyProvider: MapViewDependencyProviderProtocol {
+    func makeMetalView(frame: CGRect, device: MTLDevice?) -> MTKView {
+        MTKView(frame: frame, device: device)
+    }
+
+    func makeDisplayLink(window: UIWindow, target: Any, selector: Selector) -> DisplayLinkProtocol? {
+        window.screen.displayLink(withTarget: target, selector: selector)
+    }
+}
```
Comment (nishant-karajgikar) [inline Sources/MapboxMaps/Foundation/MapViewDependencyProvider.swift:16]: 🙇🏾 this is awesome.. We've needed to do this for a long time!

File: Tests/MapboxMapsTests/Foundation/MapViewTests.swift [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        XCTAssertEqual(displayLink.preferredFramesPerSecond, mapView.preferredFramesPerSecond.rawValue)
+
+        mapView.preferredFramesPerSecond = PreferredFPS(rawValue: 23)
+
+        XCTAssertEqual(displayLink.preferredFramesPerSecond, 23)
+    }
+
+    func testAnimatorCompletionBlocksAreRemoved() throws {
+        let firstCompletion = PendingAnimationCompletion(completion: { _ in }, animatingPosition: .end)
+        let secondCompletion = PendingAnimationCompletion(completion: { _ in }, animatingPosition: .current)
```
Comment (macdrevx) [inline Tests/MapboxMapsTests/Foundation/MapViewTests.swift:85]: Add verification that the blocks were actually invoked

File: Tests/MapboxMapsTests/Foundation/Mocks/MockMapViewDependencyProvider.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+final class MockMapViewDependencyProvider: MapViewDependencyProviderProtocol {
+    struct MakeMetalViewParams {
+        var frame: CGRect
+        var device: MTLDevice?
+    }
+    let makeMetalViewStub = Stub<MakeMetalViewParams, MockMetalView?>(defaultReturnValue: nil)
+    func makeMetalView(frame: CGRect, device: MTLDevice?) -> MTKView {
+        makeMetalViewStub.returnValueQueue.append(MockMetalView(frame: frame, device: device))
+        return makeMetalViewStub.call(with: MakeMetalViewParams(frame: frame, device: device))!
+    }
```
Comment (macdrevx) [inline Tests/MapboxMapsTests/Foundation/Mocks/MockMapViewDependencyProvider.swift:13]: Add return value generator block

**Review by nishant-karajgikar — APPROVED** [review]

**Pull request merged**